artifacts.go

```diff
@@ -99,45 +100,79 @@ func (errArtifact ErrorArtifact) ResponseObject() interface{} {
+	baseName := filepath.Base(rawContentFile)
+	tmpFile, err := ioutil.TempFile("", baseName)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer tmpFile.Close()
+	gzipLogWriter := gzip.NewWriter(tmpFile)
+	gzipLogWriter.Name = baseName
+	rawContent, err := os.Open(rawContentFile)
+	if err != nil {
+		return "", err
+	}
+	defer rawContent.Close()
+	io.Copy(gzipLogWriter, rawContent)
+	gzipLogWriter.Close()
+	return tmpFile.Name(), nil
+}
+
+func (artifact S3Artifact) ProcessResponse(resp interface{}) (err error) {
 	response := resp.(*queue.S3ArtifactResponse)
+	rawContentFile := filepath.Join(TaskUser.HomeDir, artifact.Base().CanonicalPath)
+
+	// if using Content-Encoding: gzip then need to gzip content...
+	transferContentFile := rawContentFile
+	if artifact.ContentEncoding == "gzip" {
+		transferContentFile, err = gzipCompressFile(rawContentFile)
+		if err != nil {
+			return err
+		}
+		defer os.Remove(transferContentFile)
+	}
+
+	// perform http PUT to upload to S3...
 	httpClient := &http.Client{}
 	httpCall := func() (*http.Response, error, error) {
-		// instead of using fileReader, read it into memory and then use a
-		// bytes.Reader since then http.NewRequest will properly set
-		// Content-Length header for us, which is needed by the API we call
-		fileReader, err := os.Open(filepath.Join(TaskUser.HomeDir, artifact.Base().CanonicalPath))
-		requestPayload, err := ioutil.ReadAll(fileReader)
+		transferContent, err := os.Open(transferContentFile)
 		if err != nil {
 			return nil, nil, err
 		}
-		defer fileReader.Close()
-		bytesReader := bytes.NewReader(requestPayload)
-		// http.NewRequest automatically sets Content-Length correctly for bytes.Reader
-		httpRequest, err := http.NewRequest("PUT", response.PutURL, bytesReader)
+		defer transferContent.Close()
+		transferContentFileInfo, err := transferContent.Stat()
+		if err != nil {
+			return nil, nil, err
+		}
+		transferContentLength := transferContentFileInfo.Size()
+
+		httpRequest, err := http.NewRequest("PUT", response.PutURL, transferContent)
 		if err != nil {
 			return nil, nil, err
 		}
-		log.Printf("Content-Type in S3 PUT request: %v", artifact.MimeType)
 		httpRequest.Header.Set("Content-Type", artifact.MimeType)
-		if artifact.ContentEncoding != "" {
-			httpRequest.Header.Set("Content-Encoding", artifact.ContentEncoding)
-			log.Printf("Content-Encoding in put request: %v", artifact.ContentEncoding)
+		httpRequest.ContentLength = transferContentLength
+		if artifact.ContentEncoding == "gzip" {
+			httpRequest.Header.Set("Content-Encoding", "gzip")
 		}
-		// request body could be a) binary and b) massive, so don't show it...
-		requestFull, dumpError := httputil.DumpRequestOut(httpRequest, false)
+		requestHeaders, dumpError := httputil.DumpRequestOut(httpRequest, false)
 		if dumpError != nil {
 			log.Println("Could not dump request, never mind...")
 		} else {
 			log.Println("Request")
-			log.Println(string(requestFull))
+			log.Println(string(requestHeaders))
```

Comment:
It contains just the http headers, rather than the full http payload (headers + body). I felt the old name did not reflect that very well, so renamed it.